File: shared/voice/tests/doubao_auth.cpp
#include "DoubaoAuth.h"
#include <cassert>

int main() {
  using msime::voice::doubao_auth_headers;
  const auto api = doubao_auth_headers("api_key", "stale-app", "synthetic-token", "fixture-resource");
  assert(api);
  assert(api->find("x-api-key: synthetic-token\r\n") != std::string::npos);
  assert(api->find("x-api-app-key:") == std::string::npos);
  assert(api->find("stale-app") == std::string::npos);
  const auto legacy = doubao_auth_headers("legacy", "synthetic-app", "synthetic-token", "fixture-resource");
  assert(legacy);
  assert(legacy->find("x-api-app-key: synthetic-app\r\n") != std::string::npos);
  assert(legacy->find("x-api-access-key: synthetic-token\r\n") != std::string::npos);
  assert(legacy->find("x-api-key:") == std::string::npos);
  assert(!doubao_auth_headers("legacy", "", "synthetic-token", "fixture-resource"));
  assert(!doubao_auth_headers("invalid", "synthetic-app", "synthetic-token", "fixture-resource"));
  assert(!doubao_auth_headers("api_key", "", "injected\r\nheader", "fixture-resource"));
  assert(doubao_auth_headers("", "synthetic-app", "synthetic-token", "fixture-resource"));
}

File: crates/client-core/src/doubao_auth.rs
//! Shared authentication policy for credential probes and native recognition.

fn usable(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 8192
        && !value.chars().any(char::is_control)
        && !value.starts_with('<')
        && !value.chars().all(|c| c == '*')
}

/// Produces sensitive request headers; callers must not log or persist them.
/// An absent historical mode infers legacy auth from a usable App ID. Explicit
/// API-key mode always ignores stale App IDs, including masked placeholders.
pub fn headers(
    mode: &str,
    app_id: &str,
    token: &str,
    resource_id: &str,
) -> Option<Vec<(&'static str, String)>> {
    let (app_id, token, resource_id) = (app_id.trim(), token.trim(), resource_id.trim());
    let legacy = match mode.trim() {
        "api_key" => false,
        "legacy" => true,
        "" => usable(app_id),
        _ => return None,
    };
    if !usable(token) || !usable(resource_id) || (legacy && !usable(app_id)) {
        return None;
    }
    let mut headers = vec![
        ("x-api-resource-id", resource_id.to_owned()),
        ("x-api-request-id", uuid::Uuid::new_v4().to_string()),
    ];
    if legacy {
        headers.push(("x-api-app-key", app_id.to_owned()));
        headers.push(("x-api-access-key", token.to_owned()));
    } else {
        headers.push(("x-api-key", token.to_owned()));
    }
    Some(headers)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn explicit_mode_wins_and_legacy_documents_remain_compatible() {
        for (mode, app, legacy) in [
            ("api_key", "stale-app", false),
            ("api_key", "<stored>", false),
            ("legacy", "synthetic-app", true),
            ("", "synthetic-app", true),
            ("", "", false),
        ] {
            let result = headers(mode, app, " synthetic-token ", " fixture-resource ").unwrap();
            let get = |name| {
                result
                    .iter()
                    .find(|(key, _)| *key == name)
                    .map(|(_, value)| value.as_str())
            };
            assert_eq!(get("x-api-key").is_some(), !legacy);
            assert_eq!(get("x-api-app-key").is_some(), legacy);
            assert_eq!(get("x-api-access-key").is_some(), legacy);
            assert_eq!(get("x-api-resource-id"), Some("fixture-resource"));
            assert!(uuid::Uuid::parse_str(get("x-api-request-id").unwrap()).is_ok());
        }
    }

    #[test]
    fn malformed_or_missing_auth_is_rejected_without_echoing_secrets() {
        assert!(headers("unknown", "app", "synthetic-token", "resource").is_none());
        for value in ["", "***", "<stored>", "injected\r\nheader"] {
            assert!(headers("legacy", value, "synthetic-token", "resource").is_none());
            assert!(headers("api_key", "ignored", value, "resource").is_none());
            assert!(headers("api_key", "ignored", "synthetic-token", value).is_none());
        }
        assert!(headers("api_key", "", &"x".repeat(8193), "resource").is_none());
    }
}
